Do not delete `?next` query param when redirected from AuthMessage to SignInPage

apurva-modi opened the pull request:
Passed the param next to the route whereToNext

…

Step 1: Open the link in incognito mode.
Go to https://kolibri-beta.learningequality.org/en/coach/#/0ca0280c91ccb9a880d922914226557f/plan/lessons/a7666d7f39eb34f13965ffa2e443bdee 

or 

http://127.0.0.1:8000/en/coach/#/cf75dbe3de5a09a2fc9e45702cd6a84e/plan/lessons/0a41af5b21df81371f45d8f05596b3a3

Step 2: Open use the same link to and open it in normal mode.
and you will see something like 

http://127.0.0.1:8000/en/user/#/signin?next=http%3A%2F%2F127.0.0.1%3A8000%2Fen%2Fcoach%2F%23%2Fcf75dbe3de5a09a2fc9e45702cd6a84e%2Fplan%2Flessons%2F0a41af5b21df81371f45d8f05596b3a3

where `?next` param is preserved.

Step 3: Pick a facility, sign-in, etc. and it will redirect you to the STEP 1's resource.


…

### References
FIx for #7896
…

----

### Contributor Checklist


PR process:

- [x] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

<!--
### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md
-->

jonboiser commented on kolibri/plugins/user/assets/src/routes.js:
```diff
@@ -52,6 +52,9 @@ export default [
         if (store.getters.facilities.length > 1 && !store.state.facilityId) {
           // Go to FacilitySelect with whereToNext => SignUpPage
           const whereToNext = router.getRoute(ComponentMap.SIGN_IN);
+          if ((to.query || {}).next) {
```
```suggestion
          if (to.query.next) {
```

I'm pretty sure `to.query` is always defined. Did you get errors if you leave out the default `{}`?

jonboiser commented on kolibri/plugins/user/assets/src/routes.js:
```diff
@@ -52,6 +52,9 @@ export default [
           const whereToNext = router.getRoute(ComponentMap.SIGN_IN);
+          if (to.query.next) {
+            whereToNext.query.next = to.query.next;
+          }
```
This needs to go into the `query` of `route` below. When I get redirected, I just go to this URL, with no `?next` param

http://localhost:8000/en/user/#/facilities

```
          const route = {
            ...router.getRoute(ComponentMap.FACILITY_SELECT),
            params: { whereToNext },
			query: { to.query.next } (only if to.query.next is provided)
          };
```

jonboiser replied:
Did you create a second facility when you were testing this? If you have more than 1 facility, you will go to the `/facilities` route instead of the `/sign-in` route. Run this command in a second terminal to create a new facility (replace `user:pass` with your own credentials).

```
curl -X "POST" "http://localhost:8000/api/auth/facility/" \
     -H 'Content-Type: application/json; charset=utf-8' \
     -u 'user:pass' \
     -d $'{
  "name": "Second Facility"
}'

```

jonboiser reviewed (changes requested):
The `next` param needs to go to the `query` to fix missing `next` param on the `/facilities` route. When testing, you were only looking at the `/sign-in` route, which was not broken previously.

apurva-modi commented:
> The `next` param needs to go to the `query` to fix missing `next` param on the `/facilities` route. When testing, you were only looking at the `/sign-in` route, which was not broken previously.

The whereToNext contains the information on where to be routed after the facility selection. So even though the facility selection page does not have the next param, the route after facility selection will have. 

Do you want me to change it so that even the facility selection page has the next param?

jonboiser commented:
> Do you want me to change it so that even the facility selection page has the next param?

Yes, that's the only way you'll get redirected back to the link in Step 1. The sequence of URLS needs to be

1. http://127.0.0.1:8000/en/coach/#/cf75dbe3de5a09a2fc9e45702cd6a84e/plan/lessons/0a41af5b21df81371f45d8f05596b3a3
2. http://127.0.0.1:8000/en/user/#/facilities?next=[encoded URL in Step 1]
3. Pick a Facility
3. http://127.0.0.1:8000/en/user/#/signin?next=[encoded URL in Step 1]
4. Provide credentials
5. Get authenticated, and be redirected to the URL in Step 1

apurva-modi committed:
Maintain next param across facility-sign_in redirection

The next param is provided with the route when the redirection occurs. This will make sure that the final page that is shown to the user after signin is valid.

apurva-modi commented:
> > Do you want me to change it so that even the facility selection page has the next param?
> 
> Yes, that's the only way you'll get redirected back to the link in Step 1. The sequence of URLS needs to be
> 
> 1. http://127.0.0.1:8000/en/coach/#/cf75dbe3de5a09a2fc9e45702cd6a84e/plan/lessons/0a41af5b21df81371f45d8f05596b3a3
> 2. [http://127.0.0.1:8000/en/user/#/facilities?next=[encoded](http://127.0.0.1:8000/en/user/#/facilities?next=%5Bencoded) URL in Step 1]
> 3. Pick a Facility
> 4. [http://127.0.0.1:8000/en/user/#/signin?next=[encoded](http://127.0.0.1:8000/en/user/#/signin?next=%5Bencoded) URL in Step 1]
> 5. Provide credentials
> 6. Get authenticated, and be redirected to the URL in Step 1

I have made the changes as per your suggestions. Please review.

jonboiser commented on kolibri/plugins/user/assets/src/views/FacilitySelect.vue:
```diff
@@ -106,10 +106,14 @@
     },
     methods: {
       setFacility(facilityId) {
+        let whereToNext = { ...this.whereToNext };
```
You can make this a const, since it's not reassigned 
```suggestion
        const whereToNext = { ...this.whereToNext };
```

jonboiser commented on kolibri/plugins/user/assets/src/views/SignInPage/index.vue:
```diff
@@ -211,7 +211,11 @@
+        if (this.nextParam) {
+          query = { next: this.nextParam };
+        }
+        return { ...facilityRoute, params: { whereToNext }, query: query };
```
Use object key shorthand here
```suggestion
        return { ...facilityRoute, params: { whereToNext }, query };
```

jonboiser commented on kolibri/plugins/user/assets/src/routes.js:
```diff
@@ -52,9 +52,14 @@ export default [
           const route = {
             ...router.getRoute(ComponentMap.FACILITY_SELECT),
             params: { whereToNext },
+            query: query,
```
```suggestion
            query,
```

jonboiser committed:
Apply suggestions from code review

jonboiser reviewed (approved):
Looks good!
